tfsdk/data_source_validation.go

```diff
@@ -0,0 +1,33 @@
+package tfsdk
+
+import (
+	"context"
+)
+
+// DataSourceConfigValidator describes reusable DataSource configuration validation functionality.
+type DataSourceConfigValidator interface {
+	// Description describes the validation in plain text formatting.
+	Description(context.Context) string
+
+	// MarkdownDescription describes the validation in Markdown formatting.
+	MarkdownDescription(context.Context) string
+
+	// Validate performs the validation.
+	Validate(context.Context, ValidateDataSourceConfigRequest, *ValidateDataSourceConfigResponse)
+}
+
+// DataSourceWithConfigValidators is an interface type that extends DataSource to include declarative validations.
+type DataSourceWithConfigValidators interface {
+	DataSource
+
+	// ConfigValidators returns a list of functions which will all be performed during validation.
+	ConfigValidators(context.Context) []DataSourceConfigValidator
+}
+
+// DataSourceWithValidateConfig is an interface type that extends DataSource to include imperative validation.
```

Comment:
Ideally, there'd be some description of when to use this versus the DataSourceConfigValidators.

Reply:
I hope I've covered this okay with bffb3f6d5c96ef58d0094ed78a1e9c708617a22e, please suggest better comments!